### ISSUE TITLE
Parse server replay files

### ISSUE THREAD

**Issue by TaLajans** [kotiq/wt-tools#2]
Hello,

would it be possible to improve wrpl parser to parse server replay files from https://warthunder.com/en/tournament/replay/ to get key server data and ignore all other information that could change during next update?

- Timestamp of user spawn a vehicle. 
- Timestamp of user scoring a kill/hit/crit hit in specific vehicle.
- Timestamp of user successfully capping a point.
- Timestamp of user dying in vehicle

<hr>

Right now [wrpl_unpacker_2.py ](https://github.com/kotiq/wt-tools/blob/new-format/src/wt_tools/wrpl_unpacker_2.py) can parse only local replay files and it exports map json, final result json and file called wrplu.bin.

Server replays looks like this wrplu.bin, but some times it can be split into N number of files like in this example https://warthunder.com/en/tournament/replay/type/157247827975305385
![image](https://user-images.githubusercontent.com/92247810/137772327-93e1bd3a-8582-4b75-8fbb-603c1ef88dc0.png)

### Problem
This mission end window is displayed only once in a game right after you leave a match or it ended.

![init](https://user-images.githubusercontent.com/92247810/136696486-15a94eab-32cd-4ff0-86d4-ca0ac6808607.jpg)

From this window you can get a lot of information about each of the vehicle you used in this match.
- Time spent.
- Total critical hits
- Total hits
- Total kills
- Total caps
- some other.

![time](https://user-images.githubusercontent.com/92247810/136696481-c45f8600-122c-4766-ab98-1b2a749756a7.jpg)
![crit](https://user-images.githubusercontent.com/92247810/136696483-bdc67930-2436-4800-9e7e-42a5aed1b66b.jpg)
![groundkills](https://user-images.githubusercontent.com/92247810/136696484-aa11dcb5-f62f-462b-84bd-1b6f9b2515af.jpg)
![hits](https://user-images.githubusercontent.com/92247810/136696485-aeefbd41-77a3-416c-8ffc-6b234ed90573.jpg)

**Comment by kotiq** [comment]
@TaLajans This will require in-depth knowledge of the game protocol, which I don't have right now. 
I saw a similar project for WOT created by `evido`.

**Comment by TaLajans** [comment]
I saw that WOT parser from evido too. I do hope that in near future you or someone else could get that knowledge of the game protocol and share it with everyone so we could have a similar tool for WT.

**Comment by Keksilton** [comment]
Server replays do not include useful information in the same format, so right now it's not possible.
You can check out Replay reader over on https://keksilton.github.io/WtTools.Web/
It's still limited to local replays and basic data, but it might be more convenient, I guess.

**Comment by TaLajans** [comment]
Thank you for this replay reader @Keksilton, but as of now it only gives nice scoreboard visualization, but you already could extract it using this [wrpl_unpacker_ng.py](https://github.com/kotiq/wt-tools/blob/new-format/src/wt_tools/wrpl_unpacker_ng.py). 

It looks like a file as you call `wrplu.bin` is a bit more like server replays, but as @kotiq mention you would need in-depth knowledge of game protocol to parse it.

**Comment by prandleman** [comment]
@TaLajans 

I've been working on a clog parser that can get you 

- [x] Timestamp of user spawn a vehicle.
- [x] Timestamp of user scoring a kill/hit/crit hit in specific vehicle.  ------ **yes but AFAIK only hits output to battle/damage log**
- [ ] Timestamp of user successfully capping a point. ------------ **maybe?**
- [x] Timestamp of user dying in vehicle

It was designed as a real-time parser to get details such as level binary, mission blk, team information etc. I haven't directly looked for information on point capture, but since it HUD messages are output to the log one might be able to see who the 
 point capture award (is there an award for one capture or only multiple, not sure) had been awarded to. I'm not sure if which point was captured is important to you, but I'm sure the icon changes on the web map so yeah..
 
 Only issue I see is that the replay file would have to be played back to get the information to appear in the clog. If participants sent them or they were uploaded to server the parse would be almost instant with timestamps for all the events. If you were curious the main way to tell if player has spawned or died is not by the web map json request api (cause that's a bit funky with how it's state machine works and the valid keys for indicators and state) but buy watching the camera controller output log.

**Comment by kotiq** [comment]
@prandleman 
Processing the binary log in real time can lead to a gaming advantage, which formally violates the user agreement.
For this reason, I did not finish building the state machine for events from the log. 
I would not like ordinary users to be banned.
I hope the parser from `thundertac` hasn't gone too far yet.

**Comment by tarkis** [comment]
@prandleman, as I understand you would need to launch server or player replay via game to generate this clog?

While I would like to start with something small like spawn, death, kill and mission outcome (win/lose). To have timestamps would be great if available, and you are not getting any award for capping first base only from second one and onwards.

I'm curious on how optimize it would be run. You are not allowed to run multiple instances of the game and because of this it could be the biggest bottleneck. Some fights take 10 - 30 minutes to finish. So if I would get 10 replays or more it could take few hours just to replay them. Is there a way to run WT in background like headless process to generate clog?

The ideal solution would be as @kotiq has some the scripts that just takes file as parameter and parses it. Does any part from clog and server replay binary matches that could be used?

**Comment by TaLajans** [comment]
@prandleman, as I understand you would need to launch server or player replay via game to generate this clog?

While I would like to start with something small like spawn, death, kill and mission outcome (win/lose). To have timestamps would be great if available, and you are not getting any award for capping first base only from second one and onwards.

I'm curious on how optimize it would be run. You are not allowed to run multiple instances of the game and because of this it could be the biggest bottleneck. Some fights take 10 - 30 minutes to finish. So if I would get 10 replays or more it could take few hours just to replay them. Is there a way to run WT in background like headless process to generate clog?

The ideal solution would be as @kotiq has some the scripts that just takes file as parameter and parses it. Does any part from clog and server replay binary matches that could be used?

**Comment by prandleman** [comment]
@TaLajans Yeah the reason I haven't put my parser up is that I has the xor key hard coded into the file and can be used to get an advantage with one line from the log (with a bit of work I might add). I think it's application is best with CDK power users that accept the risk and the real-time nature of it is just doing regex searches which could be liken to parsing the clog and doing a ctrl+f find lol.

Let's revisit this idea but I will need some time to hunt for my original work on parsing the damage log given by the webapi, but you should know the timestamps would be spotty. Maybe if the wrpl parsed files have a start time (I can't recall atm), but yeah I wanna stay away from the clog parser idea. Here is an example of what I can do with the webapi (red is regex caught, grey is missed) this version is not the best iteration and doesn't have any sort of timestamp. 

If you wanna discuss more talk to me in discord cause I don't want to have a chain of messages in the issue section: bit.ly/ttaccord look for the discord server creator

![image](https://user-images.githubusercontent.com/94949194/149631513-6480ebb4-2a3f-46b7-a864-05d7a9a3fba4.png)

these all come from public webapi ("http://localhost:8111/hudmsg?lastEvt=0&lastDmg=")

**Comment by prandleman** [comment]
@kotiq Nice connect there. I forgot I switched to my school account on github. Yeah, I contemplated it for a long time. Being very new to programming and software development I am still getting use to the lack of knowledge from the end user. So hard to get them to open a text file and jot in some data.. And yeah, the real-part was mostly to understand xor and debug mission editor trigger logic as well as 3d model / fm cdk errors, but then I realized I could do automate more for the end user, but obviously with some risk. 

I'll remove all that cause I do worry about the end user. I haven't published any of the code that does this AFAIK, but Gaijin would know I was getting the player name somehow if they ever tried it lol. 

With my current knowledge, the only thing that I'd be losing is the player name and which team they were on (which is kind of big, but AFAIK thundertac isn't used on a full team scale just 1v1 dogfights, so assigning the teams wouldn't be too much of a hassle. The wt root directory, current level.bin and anything else I needed can be acquired without parsing the clog. I guess I will revisit the idea of using the wrpl files post match.

On that, the only thing that I found that could be really used to cheat would be the FMW  lines.

**Comment by TaLajans** [comment]
If there is someone who would do such parser let me know. I'm willing to pay/donate up to 1k euro for such work.

**Comment by nilshellerhoff** [comment]
Hey @TaLajans ,
I have spent some time analyzing server replay files, you can see the efforts over at https://github.com/nilshellerhoff/warthunder-replay-parser. I'm not very experienced in anything binary though, but essentially you can see which vehicles have been involved in a battle. Also there is some stuff for automatically downloading replay files and getting vehicle info from the wiki.

PS: A lot of stuff is missing, I have e.g. not been able to get teams info (which player was on which team) or any information about battle progress. This was more about getting some matchmaker stats
